### ISSUE TITLE
Optimization Utility API

### ISSUE THREAD

**Issue by tylerflex** [flexcompute/tidy3d#926]
Writing down some thoughts on implementing an optimizer class for Tidy3D

## Considerations / Desired features

First, the general API I'm thinking of
```py
optimizer=Optimizer(
    params_start=params0,
    fname="optimizer.json",
    verbose=True,
    beta=1.0
)

for i in range(10):
    # pass objective function here, update current "params"
    # val and grad fn computed and evaluated under hood.
    optimizer.step(J, beta=0.1 * i)

# get the history from the OptimizerHistory storage class
optimizer.history.val # list of vals
optimizer.history.grad # list of grads
```

### Storage / History
1. Stores full history of val, grad, params, optimizer fields.
```py
optimizer=Optimizer(params)
for step_index in range(10):
    optimizer.step(J)

optimizer.history.val # float
optimizer.history.grad # array, or tuple of arrays
optimizer.history.params # dict
```
2. Resume optimizer from history
```py
optimizer=Optimizer(params)
for step_index in range(5):
    optimizer.step(J)
for step_index in range(5):
    optimizer.step(J)
```
3. Save optimizer to file by default to enable loading and resuming.
```py
fname="my_optimizer.json"
optimizer=Optimizer(params, fname=fname)
for step_index in range(5):
    optimizer.step(J)

# session interrupted

optimizer = Optimizer.from_file(fname)
for step_index in range(5):
    optimizer.step(J)
```

### Flexibility
4. Allow parameters to change each step arbitrarily and depending on
```py
optimizer=Optimizer(params, p1=1.0)
for step_index in range(10):
    if step_index <= 5:
         optimizer.step(J)
    else:
        optimizer.step(J, p1=2.0)
 ```
5. Allow objective function to change with each step?
```py
optimizer=Optimizer(params)
for step_index in range(10):
    if step_index <= 5:
         optimizer.step(J1)
    else:
        optimizer.step(J2)
``` 
6. Arbitrary callback function? (is this even necessary? what would be the signature?)
```py
optimizer=Optimizer(params)
for step_index in range(10):
    optimizer.step(J, callback=callback_fn)
```

### Usability
7. Log outputs and toggle verbosity (maybe also log level for the optimizer?)
```py
optimizer=Optimizer(params, verbose=True)
for step_index in range(10):
    optimizer.step(J) # logs output
```
8. Accept bounds and handle them for boundless algorithms (like adam)
```py
optimizer=Optimizer(params, verbose=True)
for step_index in range(10):
    optimizer.step(J) # logs output
```
9. Run function10. 
```py
optimizer=Optimizer(params, verbose=True)
optimizer.run(num_steps=10)
```

## Main Controversial Decisions
a. Do we allow this to be mutable? or enforce `optimizer = optimizer.step()`
b. Do we pass `J` every step? or just make it a field (note, can't serialize without pickle, but if we do pass J, we need to construct it again if the session is interrupted.)

**Comment by ianwilliamson** [comment]
If you are plugging into the JAX ecosystem, I would suggest checking out Optax (https://jaxopt.github.io/stable/) or JAXopt (https://jaxopt.github.io/stable/) instead of reinventing a new optimization API.

**Comment by ianwilliamson** [comment]
Also, for optimizer checkpointing to disk I would suggest checking out Orbax (https://github.com/google/orbax).

**Comment by tylerflex** [comment]
Thanks @ianwilliamson , yea this optimizer would accept a Jax-differentiable function `J`, would call `jax.value_and_grad(J)` under the hood, and pass this info to either a spicy / other optimizer or our own custom optimizers.

I've played with `jaxopt` a bit in the past and found it slightly lacking in customizability for what I was using it for, but probably it's time to do a deep dive into this as I probably just hadn't explored it enough. The other packages you suggest also look useful, thanks!

**Comment by tylerflex** [comment]
Fixed by #1012 , using `optax` in notebooks.
